### PR TITLE
Add support for literals in device factory mappings

### PR DESCRIPTION
**Pull request by tcalmant**
* Replaced integration tests by unit tests
* Allow the definition of constants in mapping configurations
* Allow default values in mapping configurations, used if the entry is **missing** (missing CSV column, missing entry in JSON)
* Added a `datetime.timezone` to use as fallback if there is no timezone/offset information in the time entry (UTC by default)
* Allow to indicate the locale to use to parse floats: `numbers.locale`
* Added string variable indexing support, *e.g.* `${name[0:4]}`

For example:
```json
{
  "$sensorId": "Id",
  "@provider": {
    "literal": "temperature-${sensorId}"
  },
  "data/value": "Value",
  "sensor/city": {
    "literal": "Grenoble"
  },
  "sensor/street": {
    "path": "Street",
    "default": "n/a",
    "type": "string"
  }
}
```

### CODE REVIEW

**Review by timothyjward — APPROVED** [review]

File: prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/RecordPath.java [hunk 1]
```diff
@@ -12,232 +12,126 @@
...
-                    arrayAsList.add(Array.get(value, i));
-                }
-                stream = arrayAsList.stream();
-            } else {
-                // Maybe a single item
-                return List.of(itemConverter.apply(value));
-            }
-
-            return stream.map(itemConverter).collect(Collectors.toList());
+    /**
```
Comment (timothyjward) [inline prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/RecordPath.java:25]: Copyright in this file is outdated

File: prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/ResourceLiteralMapping.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
+/*********************************************************************
+* Copyright (c) 2022 Contributors to the Eclipse Foundation.
```
Comment (timothyjward) [inline prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/ResourceLiteralMapping.java:2]:
```suggestion
* Copyright (c) 2023 Contributors to the Eclipse Foundation.
```

File: prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/ResourceRecordMapping.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/*********************************************************************
+* Copyright (c) 2022 Contributors to the Eclipse Foundation.
```
Comment (timothyjward) [inline prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/ResourceRecordMapping.java:2]:
```suggestion
* Copyright (c) 2023 Contributors to the Eclipse Foundation.
```

File: prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/dto/DeviceMappingOptionsDTO.java [hunk 4]
```diff
@@ -27,4 +27,7 @@ public class DeviceMappingOptionsDTO {
 
     @JsonProperty("format.datetime")
     public String formatDateTime;
+
```
Comment (timothyjward) [inline prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/dto/DeviceMappingOptionsDTO.java:30]: Another copyright update

File: prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/FactoryParserHandler.java [hunk 5]
```diff
@@ -12,24 +12,28 @@
 **********************************************************************/
 package org.eclipse.sensinact.gateway.southbound.device.factory.impl;
 
+import java.time.DateTimeException;
```
Comment (timothyjward) [inline prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/FactoryParserHandler.java:15]: Another copyright year change

File: prototype/southbound/device-factory/parser-csv/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/parser/csv/CsvRecord.java [hunk 6]
```diff
@@ -15,48 +15,62 @@
 import org.apache.commons.csv.CSVRecord;
 import org.eclipse.sensinact.gateway.southbound.device.factory.IDeviceMappingRecord;
 import org.eclipse.sensinact.gateway.southbound.device.factory.RecordPath;
+import org.slf4j.Logger;
```
Comment (timothyjward) [inline prototype/southbound/device-factory/parser-csv/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/parser/csv/CsvRecord.java:18]: Another copyright year

File: prototype/southbound/device-factory/parser-csv/src/test/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/CSVParserTest.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
+/*********************************************************************
+* Copyright (c) 2022 Contributors to the Eclipse Foundation.
```
Comment (timothyjward) [inline prototype/southbound/device-factory/parser-csv/src/test/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/CSVParserTest.java:2]:
```suggestion
* Copyright (c) 2023 Contributors to the Eclipse Foundation.
```

File: prototype/southbound/device-factory/parser-json/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/parser/json/JsonParser.java [hunk 8]
```diff
@@ -14,10 +14,12 @@
 
 import java.io.IOException;
 import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
```
Comment (timothyjward) [inline prototype/southbound/device-factory/parser-json/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/parser/json/JsonParser.java:17]: Another Copyright year to update

File: prototype/southbound/device-factory/parser-json/src/test/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/JSONParserTest.java [hunk 9]
```diff
@@ -1,3 +1,5 @@
+package org.eclipse.sensinact.gateway.southbound.device.factory.impl;
+
 /*********************************************************************
 * Copyright (c) 2022 Contributors to the Eclipse Foundation.
```
Comment (timothyjward) [inline prototype/southbound/device-factory/parser-json/src/test/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/JSONParserTest.java:4]:
```suggestion
* Copyright (c) 2023 Contributors to the Eclipse Foundation.
```

File: prototype/southbound/http/http-device-factory/src/main/java/org/eclipse/sensinact/gateway/southbound/http/factory/HttpDeviceFactory.java [hunk 10]
```diff
@@ -12,6 +12,7 @@
 **********************************************************************/
 package org.eclipse.sensinact.gateway.southbound.http.factory;
```
Comment (timothyjward) [inline prototype/southbound/http/http-device-factory/src/main/java/org/eclipse/sensinact/gateway/southbound/http/factory/HttpDeviceFactory.java:13]: Another copyright year

**Comment by timothyjward** [comment]
There probably needs to be some squashing in this PR. I'm not sure if it's too big to squash into a single commit, or if there's some sensible sub-grouping?

**Comment by tcalmant** [comment]
I've squashed the commits a bit and updated copyright years.
I'm editing the PR root comment to add info on additional features

**Pull request merged**